src/ILCompiler.Compiler/src/Compiler/DelegateCreationInfo.cs

```diff
@@ -0,0 +1,131 @@
+            if (!targetMethod.Signature.IsStatic)
+            {
+                // TODO: check for explicit this and don't increment if that's the case
+                paramCountTargetMethod++;
+            }
+
+            DelegateInfo delegateInfo = context.GetDelegateInfo(delegateType.GetTypeDefinition());
+            int paramCountDelegateClosed = delegateInfo.Signature.Length + 1;
+            bool closed = false;
+            if (paramCountDelegateClosed == paramCountTargetMethod)
+            {
+                closed = true;
+            }
+            else
+            {
+                Debug.Assert(paramCountDelegateClosed == paramCountTargetMethod + 1);
+            }
+
+            if (targetMethod.Signature.IsStatic)
+            {
+                MethodDesc invokeThunk;
+                if (!closed)
+                {
+                    // Open delegate to a static method
+                    invokeThunk = delegateInfo.Thunks[DelegateThunkKind.OpenStaticThunk];
+                }
+                else
+                {
+                    // Closed delegate to a static method (i.e. delegate to an extension method that locks the first parameter)
+                    invokeThunk = delegateInfo.Thunks[DelegateThunkKind.ClosedStaticThunk];
+                }
+
+                var instantiatedDelegateType = delegateType as InstantiatedType;
+                if (instantiatedDelegateType != null)
+                    invokeThunk = context.GetMethodForInstantiatedType(invokeThunk, instantiatedDelegateType);
+
+                // We use InitializeClosedStaticThunk for both because RyuJIT generates same code for both,
+                // but passes null as the first parameter for the open one.
+                return new DelegateCreationInfo(
+                    (IMethodNode)factory.MethodEntrypoint(systemDelegate.GetKnownMethod("InitializeClosedStaticThunk", null)),
+                    factory.MethodEntrypoint(targetMethod),
+                    (IMethodNode)factory.MethodEntrypoint(invokeThunk));
+            }
+            else
+            {
+                if (!closed)
+                    throw new NotImplementedException("Open instance delegates");
+
+                bool useUnboxingStub = targetMethod.OwningType.IsValueType;
+
+                return new DelegateCreationInfo(
+                    (IMethodNode)factory.MethodEntrypoint(systemDelegate.GetKnownMethod("InitializeClosedInstance", null)),
+                    factory.MethodEntrypoint(targetMethod, useUnboxingStub));
+            }
+        }
+
+        public override bool Equals(object obj)
+        {
+            var other = obj as DelegateCreationInfo;
+            return other != null && Constructor == other.Constructor
+                && Target == other.Target && Thunk == other.Thunk;
+        }
+
+        public override int GetHashCode()
+        {
+            return Constructor.Method.GetHashCode() ^ Target.MangledName.GetHashCode();
```

Comment:
`return Constructor.GetHashCode() ^ Target.GetHashCode()` may be better - more in sync with equals.